### PR TITLE
For #12894 Canceling the "Open in app?" should open the page in focus browser.

### LINKED ISSUE #12894: Canceling the "Open in app?" displays an empty page
## Prerequisites
Have Youtube, or Maps apps installed.

## Steps to reproduce
1. Go to Settings - Advanced, and enable the "Open links in apps" toggle.
2. Exit settings, and go to youtube.com, or maps.google.com.
3. On the "Open in apps?" tap on "Cancel".
4. Observe the page.

### Expected behavior
The page loads in browser, without issues.

### Actual behavior
The page is blank.

The original ticket from Focus https://github.com/mozilla-mobile/focus-android/issues/7777 
This bug was introduced with this pr https://github.com/mozilla-mobile/android-components/pull/12382

* Android device: all devices 
* App (fenix, focus, ...): focus
* App version: from 103 until present

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21732)

### PR DESCRIPTION
**Pull request by iorgamgabriel**
The original ticket from Focus https://github.com/mozilla-mobile/focus-android/issues/7777
This bug was introduced with this pr https://github.com/mozilla-mobile/android-components/pull/12382

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
Fixes #12894

### CODE REVIEW

**Comment by iorgamgabriel** [comment]
https://user-images.githubusercontent.com/8118371/194086390-afcd682c-a17a-4613-82e5-fad06ac2b16e.mp4

**Comment by iorgamgabriel** [comment]
https://user-images.githubusercontent.com/8118371/194086446-660c936b-d370-4247-940a-856e9f1ff4be.mp4

**Comment by Amejia481** [comment]
@iorgamgabriel should we add tests to verify the new behaviour? :)

**Comment by rocketsroger** [comment]
Looks like the same issue doesn't happen on Fenix side.  Will this break anything on the Fenix side? or is there something different on the Fenix side that loads the page when the user taps cancel?


https://user-images.githubusercontent.com/6579541/194169887-0b4fbd49-d5e9-475f-a8da-9a7bab6f5e9f.mp4

**Comment by rocketsroger** [comment]
With some debugging I can see that without your fix we are already calling `loadUri` when the user taps "Cancel". Not sure why it's not loadinf though.

Don't see any difference between this and Fenix private mode.  But I think this will cause an regression on Fenix side.

**Comment by rocketsroger** [comment]
@iorgamgabriel I was thinking, if this fix is a rush for Focus, I'm also OK with landing this and create an issue to investigate why this is only causing an issue on Focus but not Fenix.  Since right now there's no workaround on Focus side.

**Comment by rocketsroger** [comment]
Looked at it for a bit more.  I think if we want to make the change you have here, we'll have to update the "isDomain" check inside the `AppLinksInterceptor`.  @iorgamgabriel do you might I take a shot at this tomorrow?  Thanks

**Comment by iorgamgabriel** [comment]
On Focus is not working because interceptor.interceptsAppInitiatedRequests() 
is override interceptsAppInitiatedRequests() = true and is not entering in this
 if (interceptionResponse !is InterceptionResponse.AppIntent) {
                appRedirectUrl = ""
                lastLoadRequestUri = request.uri
            }
I can't change this in Focus because all the links  ("focus:about" ,"focus:rights", "focus:gpl","focus:licenses") will not work so this change must be in ac. 
In fenix interceptsAppInitiatedRequests()  is not override and will be the default value false. @rocketsroger

**Comment by iorgamgabriel** [comment]
I tested your scenario from the video and it's working with this change . @rocketsroger

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 1]
```diff
@@ -674,8 +674,10 @@ class GeckoEngineSession(
 
             if (interceptionResponse !is InterceptionResponse.AppIntent) {
                 appRedirectUrl = ""
-                lastLoadRequestUri = request.uri
             }
+
+            lastLoadRequestUri = request.uri
```
Comment (rocketsroger) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:679]: I suggest not to remove line 677.
We then create an issue to fix the root cause.  But for now we add this workaround for Focus.  Please let me know if this is a good compromise. 

```suggestion
            lastLoadRequestUri = request.uri
        }
            
        // TODO fix root cause: https://github.com/mozilla-mobile/android-components/issues/<TBD>
        if (interceptor != null && interceptor.interceptsAppInitiatedRequests()) {
            lastLoadRequestUri = request.uri
        }
```

**Review by rocketsroger — CHANGES_REQUESTED** [review]

**Comment by rocketsroger** [comment]
> I tested your scenario from the video and it's working with this change . @rocketsroger

Yes, but what I meant there was that it's working without this change as well.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 1 above)
Comment (iorgamgabriel) [reply components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Done

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 2]
```diff
@@ -676,6 +695,12 @@ class GeckoEngineSession(
                 appRedirectUrl = ""
                 lastLoadRequestUri = request.uri
             }
+
+            // TODO fix root cause: https://github.com/mozilla-mobile/android-components/issues/12894
```
Comment (rocketsroger) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:699]: Should we create another issue to track the fixing of the root cause?

**Review by rocketsroger — APPROVED** [review]
Looks good to me.  Only question is if we should create a new issue to track the follow up.

**Comment by jonalmeida** [comment]
@rocketsroger @iorgamgabriel tests? Wasn't this a regression? 🙂

**Comment by rocketsroger** [comment]
> @rocketsroger @iorgamgabriel tests? Wasn't this a regression? 🙂

This was not an regression.  The fix we added was a temporary workaround for the root cause.  To fix this completely we'll need to make a change in GeckoView to allow us to call `LoadUri` with a flag that bypasses interceptors.

**Pull request merged**